internal/initwd/module_install.go

```diff
@@ -936,3 +984,32 @@ func splitAddrSubdir(addr addrs.ModuleSource) (string, string) {
 		return addr.String(), ""
 	}
 }
+
+func buildModuleVersionDeprecationWarning(modVersion *response.ModuleVersion, req *configs.ModuleRequest) *hcl.Diagnostic {
+	var additionalInfo []string
+	if modVersion.Deprecation.Reason != "" {
+		additionalInfo = append(additionalInfo, modVersion.Deprecation.Reason)
+	}
+	if modVersion.Deprecation.Link != "" {
+		additionalInfo = append(additionalInfo, fmt.Sprintf("More information: %s", modVersion.Deprecation.Link))
+	}
+	var detail string
+	if len(additionalInfo) == 1 {
+		detail = additionalInfo[0]
+	} else {
+		detail = strings.Join(additionalInfo, "\n\n")
+	}
+	return &hcl.Diagnostic{
+		Severity: hcl.DiagWarning,
+		Summary:  fmt.Sprintf("Module version %s of %s is deprecated", modVersion.Version, req.Name),
```

Comment:
The name of the module... For example, "foo" in `module "foo" {` is not the thing that is specifically deprecated, and wording it this way is slightly confusing because of the word "of"... plus, it prevents default consolidation from taking place.

If your config says this:

```
module "A" {
  source = "app.staging.terraform.io/bcroft/example-module/aws"
  version = "1.0.0"
}
```

Should the message say "Module version 1.0.0 of A is deprecated" or should it say "Module version 1.0.0 of app.staging.terraform.io/bcroft/example-module/aws is deprecated" ?

The latter seems much more accurate, plus all the usage of that particular module will be consolidated by the default consolidator. Otherwise you get all the warnings in each usage since the summary varies by name

Reply:
included 💯   https://github.com/hashicorp/terraform/pull/35231/commits/e3fc6d2d30c24cd5ceff57d8b3bb6fb23cadebe5